### PR TITLE
`PwBaseWorkChain`: handle diagonalization issues

### LINKED ISSUE #742: `PwBaseWorkChain`: error handler for diagonalization issues
While running calculations (especially `nscf`, I've noticed), one often runs into the following errors:

1. Too many bands not converged
```
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
from c_bands : error # 1
too many bands are not converged
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

2. Cholesky
```
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
     Error in routine  cdiaghg (161): 
      problems computing cholesky
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
```

[2] is already parsed and raises a specific error. [1] should also be added to the list of errors for the `PwCalculation`. For both issues, I've tried switching `ELECTRONS.diagonalization` to `'cg'`, and this was successful in fixing my `nscf` calculations ( 2 cases of [1], 12 cases of [2]). 

@pietrodelugas would you say that this is a good approach to deal with these issues in general, or did I just get lucky? 🙃 

Will be doing a lot more `PwBandsWorkChain` and `PdosWorkChain`s soon, and will try adding this error handler to see how many failures it can deal with. If it performs well, I'll open a PR.

### PR DESCRIPTION
**Pull request by mbercx**
Fixes #742 

Add parsing for an error related to the diagonalization, and add a
corresponding exit code to the `PwCalculation`.

Add an error handler to deal with both this exit code as well as the
`ERROR_COMPUTING_CHOLESKY` exit code. In both cases, switching to using
conjugate gradient for the diagonalization seems to be a quite
successful approach for recovering from these failures. If the
calculation was already running conjugate gradient, the
`PwBaseWorkChain` fails with a known unrecoverable failure.

### CODE REVIEW

**Review by sphuber — COMMENTED** [review]
Looks good @mbercx , thanks. Would just at least add a test for the `PwParser` to check the error gets picked up correctly. Bonus points if you add a test for the workchain handler itself, but I think that is less critical

**Comment by mbercx** [comment]
I've been testing this in the field, and it seems to work quite nicely:

1. Cholesky issues (exit code `462`)
```
PwBaseWorkChain<213713> Finished [0] [5:results]
    ├── create_kpoints_from_distance<213820> Finished [0]
    ├── PwCalculation<213888> Finished [462]
    └── PwCalculation<215363> Finished [0]
PwBaseWorkChain<213899> Finished [0] [5:results]
    ├── create_kpoints_from_distance<213900> Finished [0]
    ├── PwCalculation<213909> Finished [462]
    └── PwCalculation<215765> Finished [0]
...
```

So far I've encountered this error 18 times in my test runs, and _every_ time switching to conjugate gradient seems to fix the problem.

2. "Too many bands not converged" (exit code `463`)
```
PwBaseWorkChain<214832> Finished [0] [5:results]
    ├── create_kpoints_from_distance<215196> Finished [0]
    ├── PwCalculation<215224> Finished [463]
    └── PwCalculation<217771> Finished [0]
PwBaseWorkChain<215379> Finished [0] [5:results]
    ├── create_kpoints_from_distance<215888> Finished [0]
    ├── PwCalculation<215926> Finished [463]
    └── PwCalculation<218431> Finished [0]
...
```

Here it also works most of the time, but there are still some failures. One run already started with `'cg'`, and yet the calculation failed:

```
PwBaseWorkChain<219924> Finished [310] [4:while_(should_run_process)(2:inspect_process)]
    └── PwCalculation<219935> Finished [463]
```

```
$ verdi process report 219924
2021-09-22 17:01:03 [45694 | REPORT]: [219924|PwBaseWorkChain|run_process]: launching PwCalculation<219935> iteration #1
2021-09-22 17:03:55 [45838 | REPORT]: [219924|PwBaseWorkChain|report_error_handled]: PwCalculation<219935> failed with exit status 463: Too many bands failed to converge during the diagonalization.
2021-09-22 17:03:55 [45839 | REPORT]: [219924|PwBaseWorkChain|report_error_handled]: Action taken: found diagonalization issues but already running with conjugate gradient algorithm, aborting...
2021-09-22 17:03:55 [45840 | REPORT]: [219924|PwBaseWorkChain|inspect_process]: PwCalculation<219935> failed but a handler detected an unrecoverable problem, aborting
2021-09-22 17:03:55 [45841 | REPORT]: [219924|PwBaseWorkChain|on_terminated]: remote folders will not be cleaned
```

**Comment by mbercx** [comment]
> Looks good @mbercx , thanks. Would just at least add a test for the `PwParser` to check the error gets picked up correctly. Bonus points if you add a test for the workchain handler itself, but I think that is less critical

Indeed, still have to fix the tests, on it! Will see how complex it is to get those bonus points. 😎

**Review by ramirezfranciscof — COMMENTED** [review]
I probably shouldn't be approving this since I never done an error handler before (so I'm more learning how it works by looking at this rather than actually checking if it works), but the general logic looks fine to me.

**Comment by mbercx** [comment]
> I probably shouldn't be approving this since I never done an error handler before (so I'm more learning how it works by looking at this rather than actually checking if it works), but the general logic looks fine to me.

You can always give it a spin for one of your problematic systems if you'd like. ^^ I've tested it in the field, though, and it seems to work fine! Just going to compare with `ParO` instead of `cg` since that should be much faster [according to Pietro](https://github.com/aiidateam/aiida-quantumespresso/issues/742#issuecomment-925141664).

**Comment by sphuber** [comment]
> Just going to compare with `ParO` instead of `cg` since that should be much faster [according to Pietro](https://github.com/aiidateam/aiida-quantumespresso/issues/742#issuecomment-925141664).

Note that [ParO wasn't added until QE v6.6](https://github.com/QEF/q-e/releases/tag/qe-6.6), so depending on our new version support protocol we may not yet start to rely on it existing. Making it conditional would be tricky because there is no official way to determine what code version will be used.

**Comment by mbercx** [comment]
> > Just going to compare with `ParO` instead of `cg` since that should be much faster [according to Pietro](https://github.com/aiidateam/aiida-quantumespresso/issues/742#issuecomment-925141664).
> 
> Note that [ParO wasn't added until QE v6.6](https://github.com/QEF/q-e/releases/tag/qe-6.6), so depending on our new version support protocol we may not yet start to rely on it existing. Making it conditional would be tricky because there is no official way to determine what code version will be used.

Good point! Thanks for looking into this. Re the version of the code: we should really move to adding this to the `Code` somehow. Can we perhaps automatically add the code version to the extras of the `Code` at the start of the parsing?

**Comment by mbercx** [comment]
@sphuber @ramirezfranciscof since we currently still support all v6.X QE versions, and have no way to determine the version of the code run before parsing, I think we can go ahead and merge this with `'cg'`. It's better to be able to handle this error with a slow method than not handle it at all.

Will open another issue to discuss how we can deal with code version-dependent logic and add a note there that we should update this error handler when we figure this out.

**Comment by mbercx** [comment]
@sphuber I got the bonus points, but had to comment out the `test_handle_known_unrecoverable_failure` test since we now have no exit codes that need to get handled by this error handler. Not sure if this is fine?

File: tests/workflows/pw/test_base.py
Comment (sphuber) [inline tests/workflows/pw/test_base.py:112]:
```suggestion
pytest.mark.skip('Reactivate once we have an unrecoverable failure once again.')
 def test_handle_known_unrecoverable_failure(generate_workchain_pw):
     """Test `PwBaseWorkChain.handle_known_unrecoverable_failure`."""
     process = generate_workchain_pw(exit_code=PwCalculation.exit_codes.<INSERT EXIT CODE HERE>)
     process.setup()

     result = process.handle_known_unrecoverable_failure(process.ctx.children[-1])
     assert isinstance(result, ProcessHandlerReport)
     assert result.do_break
     assert result.exit_code == PwBaseWorkChain.exit_codes.ERROR_KNOWN_UNRECOVERABLE_FAILURE

     result = process.inspect_process()
     assert result == PwBaseWorkChain.exit_codes.ERROR_KNOWN_UNRECOVERABLE_FAILURE
```

File: tests/workflows/pw/test_base.py
Comment (sphuber) [inline tests/workflows/pw/test_base.py:134]: wait, why is this called a second time?

File: aiida_quantumespresso/workflows/pw/base.py
Comment (sphuber) [inline aiida_quantumespresso/workflows/pw/base.py:532]:
```suggestion
```

File: aiida_quantumespresso/workflows/pw/base.py
Comment (sphuber) [inline aiida_quantumespresso/workflows/pw/base.py:540]:
```suggestion
```

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @mbercx . Couple of small things but should be good to merge after that

File: tests/workflows/pw/test_base.py
Comment (sphuber) [reply tests/workflows/pw/test_base.py]: I see, but that means we still have an `ERROR_KNOWN_UNRECOVERABLE_FAILURE` pathway and you don't have to actually skip the test that checks that path, as you do now. Anyway, this path is checked in the test discussed in this comment, so it is ok to leave for now.

**Review by sphuber — APPROVED** [review]

**Pull request merged**
